lib/mod_movements/handler.js

```diff
@@ -0,0 +1,170 @@
+const { Movement, Action } = require('./movement')
+const Vec3 = require('vec3').Vec3
+
+function deltaToVec (delta) {
+  return new Vec3(delta[0], delta[1], delta[2])
+}
+
+function radialSymmetry (delta) {
+  return [
+    [delta[0], delta[1], delta[2]],
+    [-delta[2], delta[1], delta[0]],
+    [-delta[0], delta[1], -delta[2]],
+    [delta[2], delta[1], -delta[0]]
+  ]
+}
+
+/**
+ * A parser for reading movement plugin files to generate available
+ * movement lists while pathfinding.
+ */
+class ModularMovementsHandler {
+  /**
+   * Creates a new modular movement handler instance.
+   *
+   * @param {boolean} loadDefault - Whether or not to load the default
+   * movement plugin.
+   */
+  constructor (loadDefault = true) {
+    this.movementPlugins = []
+    this.settings = require('../../movements/settings.json')
+
+    if (loadDefault) {
+      this.loadMovements(require('../../movements/defaultMovements.json'))
+    }
+  }
+
+  /**
+   * Loads a new movement plugin to this handler from a Json file.
+   *
+   * @param {*} movements - The movement Json file to load movement
+   * data from.
+   */
+  loadMovements (movements) {
+    this.movementPlugins.push(movements)
+  }
+
+  /**
+   * Gets all available movements (or neighbor nodes) which can be preformed
+   * from the given node.
+   *
+   * @param {*} node - The node to handle.
+   * @param {*} world - The virtual world state.
+   */
+  getNeighbors (node, world) {
+    const neighbors = []
+    for (const plugin of this.movementPlugins) {
+      for (const movement of plugin.movements) {
+        if (movement.symmetry === 'radial') {
+          for (let i = 0; i < 4; i++) {
+            const m = this._handleMovement(movement, i, node, world)
+            if (m) neighbors.push(m)
+          }
+        } else {
+          const m = this._handleMovement(movement, 0, node, world)
+          if (m) neighbors.push(m)
+        }
+      }
+    }
+
+    return neighbors
+  }
+
+  /**
+   * Returns a new movement node if this current movement is considered
+   * valid.
+   *
+   * @param {*} movement - The raw movement to test for validity.
+   * @param {*} node - The node currently being handled.
+   * @param {*} world - The virtual world state.
+   */
+  _handleMovement (movement, direction, node, world) {
+    const hash = this._generateHash(movement.hash, node.position)
+    const m = new Movement(node.position, hash)
+
+    for (const action of movement.actions) {
+      const type = action.type
+      const delta = deltaToVec(radialSymmetry(action.delta)[direction])
+      const requirements = action.requires || []
+
+      if (!this._meetsRequirements(node, world, requirements)) continue
+
+      const block = world.virtualBlockAt(node, node.position + delta)
+      const a = new Action(type, block.position)
+
+      if (type === 'mine') {
+        if (!this._canMine(block)) continue
+
+        a.tool = 'hand' // TODO Determine best tool
```

Comment:
This is better left as a runtime decision. It doesn't impact which path is better, you are only interested to know whether or not you can break the block, you don't need to store the tool you need. Memory is precious and checking the inventory for best tool is fast enough to be redone when walking the path.

Reply:
While I agree that should be the default option, I think considering this information to be considered at solve time is also useful if the developer decides to turn that on. It can be much easier to travel a few extra blocks and tunnel through a dirt wall than try and mine through an Obsidian wall with your hand.

Reply:
It is already considered as part of the cost of the movement. But you don't need to store the tool name.